packages/common/src/repo/repo.ts

```diff
@@ -0,0 +1,455 @@
+    this.cid = params.cid
+    this.authStore = params.authStore
+  }
+
+  static async create(
+    blockstore: IpldStore,
+    did: string,
+    authStore: auth.AuthStore,
+  ): Promise<Repo> {
+    const foundUcan = await authStore.findUcan(auth.maintenanceCap(did))
+    if (foundUcan === null) {
+      throw new Error(`No valid Ucan for creating repo`)
+    }
+    const tokenCid = await blockstore.put(auth.encodeUcan(foundUcan))
+
+    const data = await MST.create(blockstore)
+    const dataCid = await data.save()
+
+    const root: RepoRoot = {
+      did,
+      prev: null,
+      auth_token: tokenCid,
+      data: dataCid,
+    }
+
+    const rootCid = await blockstore.put(root)
+    const commit: Commit = {
+      root: rootCid,
+      sig: await authStore.sign(rootCid.bytes),
+    }
+
+    const cid = await blockstore.put(commit)
+
+    return new Repo({
+      blockstore,
+      data,
+      commit,
+      root,
+      cid,
+      authStore,
+    })
+  }
+
+  static async load(blockstore: IpldStore, cid: CID, authStore?: AuthStore) {
+    const commit = await blockstore.get(cid, schema.commit)
+    const root = await blockstore.get(commit.root, schema.repoRoot)
+    const data = await MST.load(blockstore, root.data)
+    return new Repo({
+      blockstore,
+      data,
+      commit,
+      root,
+      cid,
+      authStore: authStore || null,
+    })
+  }
+
+  static async fromCarFile(
+    buf: Uint8Array,
+    blockstore: IpldStore,
+    verifyAuthority = true,
+    authStore?: auth.AuthStore,
+  ) {
+    const car = await CarReader.fromBytes(buf)
+
+    const roots = await car.getRoots()
+    if (roots.length !== 1) {
+      throw new Error(`Expected one root, got ${roots.length}`)
+    }
+    const root = roots[0]
+
+    for await (const block of car.blocks()) {
+      await blockstore.putBytes(block.cid, block.bytes)
+    }
+
+    const repo = await Repo.load(blockstore, root, authStore)
+    if (verifyAuthority) {
+      await repo.verifySetOfUpdates(null, repo.cid)
+    }
+    return repo
+  }
+
+  did(): string {
+    return this.root.did
+  }
+
+  getCollection(name: string): Collection {
+    if (name.length > 256) {
+      throw new Error(
+        `Collection names may not be longer than 256 chars: ${name}`,
+      )
+    }
+    const parts = name.split('/')
+    if (parts.length > 2) {
+      throw new Error(
+        `Only one level of namespacing allowed in collection names: ${name}`,
+      )
+    } else if (parts.length < 2) {
+      throw new Error(
+        `Expected at least one level of namespacing in collection name: ${name}`,
+      )
+    }
+    return new Collection(this, parts[0], parts[1])
+  }
+
+  // The repo is mutable & things can change while you perform an operation
+  // Ensure that the root of the repo has not changed so that you don't get local branching
+  async safeCommit(
+    mutation: (data: DataStore) => Promise<DataStore>,
+  ): Promise<void> {
+    if (this.authStore === null) {
```

Comment:
Do you need to wrap this in a mutex? Line 149 applies the update to `this.data` but it's not updated until after a lot of async at 169, which I suspect could lead two `safeCommit()` calls to clobber each other.

Reply:
I generally try to stay clear of mutexes. The `DataStore` is immutable so the mutation doesn't mutate the passed in object at all, it just returns a new object with the mutation applied. So it does all the async work but doesn't mutate the _actual repo_ until it's done the check to ensure it's in the same state it thought it was.

Remembering that I meant to write an async test for this to make sure `safeCommit` is working so i'll add that in

Reply:
> The DataStore is immutable so the mutation doesn't mutate the passed in object at all, it just returns a new object with the mutation applied.

I could be wrong here -- but I think that's actually the problem. Two safeCommits running at the same time will apply to the same starting state rather than running sequentially, causing one to clobber the other.

Reply:
Turns out I missed the commit guard at https://github.com/bluesky-social/adx/pull/158/files#diff-167949500561ff7a2113519f05ddcd4d6005dac965e67c9b292116d617b6e0d0R165-R167, this is fine